presto-spi/src/main/java/com/facebook/presto/spi/function/ComplexTypeFunctionDescriptor.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.function;
+
+import com.facebook.presto.common.Subfield;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.facebook.presto.common.Subfield.allSubscripts;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptySet;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableSet;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Contains properties that describe how the function operates on Map or Array inputs.
+ */
+public class ComplexTypeFunctionDescriptor
+{
+    public static final ComplexTypeFunctionDescriptor DEFAULT = new ComplexTypeFunctionDescriptor(
+            true,
+            false,
+            true,
+            emptyList(),
+            Optional.of(emptySet()),
+            Optional.empty());
+
+    /**
+     * Indicates whether the function accessing subfields.
+     */
+    private final boolean isAccessingInputValues;
+
+    /**
+     * Indicates whether the semantic of the function transforms the shape of the input.
+     * If function does not transform the shape of the input, then all collected subfields accessed from lambdas from outer calls could be passed to the input column or next
+     * nested call. If function does change the input then the transformation could be supplied in <code>outputToInputTransformationFunction</code>.
+     * Example. For the table <code>CREATE TABLE (events array(row(id BIGINT, data VARCHAR)))</code>, let's consider the query
+     * <code>SELECT FILTER(TRANSFORM(events, x -> CAST(x AS ROW(key BIGINT, value VARCHAR), y -> y.key > 1000)))</code>.
+     * Here, the subfield <code>key</code> was captured in the function FILTER. And the function TRANSFORM transforms the input that make it impossible to pass the accessed
+     * subfield <code>key</code> from outer call FILTER without transforming the name of the subfield from <code>key</code> to <code>id</code>.
+     */
+    private final boolean isTransformingInput;
```

Comment:
similar to the discussion about isAccessingInputValues, I think we can remove this field, and instead have a combo of 
isReturningInputValues() and outputToInputTransformationFunction().  (a function is transforming input if it's returning input values and some transformation function is provided.  if not transformation function is provided, it's either not transforming input, or we don't know how so it doesn't matter).

Reply:
>if not transformation function is provided, it's either not transforming input, or we don't know how so it doesn't matter)
It actually does matter. If function is not transforming input values then we can pass accessed subfields further to the inner function. If or we don't know how it transforms the input, then we must include all subfields.

Reply:
after much consideration, I managed to remove `isReturningInputValues`. I did it in a separate commit as I do not know yet whether this approach is better.